File: deploy.py
import json, re
from zipfile import ZipFile

F_USERSCRIPT_HEADER0 = "userscript_header.js"
INLINE_VAR_REGEX = re.compile('"(CONFIG|MANIFEST):(\w+)"')
ZIPFILE0 = "annas-archive_v{}.zip"

def deploy_addon(manifest):
    version = manifest["version"]
    files0 = [
        "manifest.json",
        manifest["content_scripts"][0]["js"][0],
        manifest["web_accessible_resources"][0]
    ]

    zipfile0 = ZIPFILE0.format(version)
    with ZipFile(zipfile0, "w") as zipfile:
       for file0 in files0:
           zipfile.write(file0)

    print("Deployed add-on: " + zipfile0)

def deploy_userscript(manifest):
    config = {
        "match": manifest["content_scripts"][0]["matches"][0]
    }

    f_js0 = manifest["content_scripts"][0]["js"][0]
    with open(f_js0, "r") as f_js:
        js = f_js.read()

    with open(F_USERSCRIPT_HEADER0, "r") as f_userscript_header:
        userscript_header = f_userscript_header.read()

    matches = INLINE_VAR_REGEX.findall(userscript_header)
    for location, var in matches:
        str_a = '"{}:{}"'.format(location, var)
        if location == "MANIFEST":
            str_b = manifest[var]
        elif location == "CONFIG":
            str_b = config[var]

        userscript_header = userscript_header.replace(str_a, str_b)

    js = userscript_header + "\n" + js
    f_js0 = f_js0[:-3] + ".user.js"
    with open(f_js0, "w") as f_js:
        f_js.write(js)

    print("Deployed userscript: " + f_js0)

def main():
    with open("manifest.json", "r") as f_manifest:
        manifest = json.load(f_manifest)

    deploy_addon(manifest)
    deploy_userscript(manifest)

if __name__ == "__main__":
    main()

File: deploy.js
const fs = require("fs")
const JSZip = require("jszip")

// filenames: in
const PACKAGE0 = "./package.json"
const JS_SKEL0 = "./rsc/annas_goodreads_skel.js"
const MANIFEST_SKEL0 = "./rsc/manifest_skel.json"
const ADDON_HEADER0 = "./rsc/addon_header.js"
const USERSCRIPT_HEADER0 = "./rsc/userscript_header.js"
const ICON0 = "./addon/annas-archive-favicon.png"

// filenames: out
const USERSCRIPT0 = "./annas_goodreads.user.js"
const ADDON_JS0 = "./addon/annas_goodreads.js"
const MANIFEST0 = "./addon/manifest.json"
const ARCHIVE0 = "annas-goodreads_v{}.zip"

const INLINE_VAR_REGEX = /"([A-Z]+):(\w+)"/g

// generates the addon text
function generateAddon(skel, config) {
  skel = skel.replace(/"USERSCRIPT:\w+"\n/g, "")
  return resolveAllVars(skel, config)
}

// generates the manifest text
function generateManifest(skel, config) {
  return resolveAllVars(skel, config, true)
}

// resolves inline vars in the userscript headers and seperates
// actual headers from definitions
function parseUserscriptHeader(skel, config) {
  let resolvedVars = resolveAllVars(skel, config)
  return resolvedVars.split("\"SPLIT\"")
}

// generates the userscript text
function generateUserscript(skel, config) {
  skel = skel.replace(/"ADDON:\w+"\n/g, "")
  return resolveAllVars(skel, config)
}

// resolves a single variable
function resolveVar(match, config) {
  let namespace = match[1]
  let variable = match[2]
  return config[namespace.toLowerCase()][variable]
}

// resolves every variable and replaces their occurences in skel
function resolveAllVars(skel, config, quotes = false) {
  const matches = [...skel.matchAll(INLINE_VAR_REGEX)] 
  matches.forEach(match => {
    let result = resolveVar(match, config)
    if (quotes) result = "\"" + result + "\""
    skel = skel.replace(match[0], result)
  })
  return skel
}

// ./path/to/file.js -> file.js
function strippedFileName(path) {
  return path.match(/[^/]+\.[^/]+$/gm)[0]
}

function main() {
  console.log("Generating files...")
  let manifest = require(MANIFEST_SKEL0)
  const packagejson = require(PACKAGE0)

  let jsSkel = fs.readFileSync(JS_SKEL0, "utf-8")
  let manifestSkel = fs.readFileSync(MANIFEST_SKEL0, "utf-8")
  let addonHeader = fs.readFileSync(ADDON_HEADER0, "utf-8")
  let userscriptHeader = fs.readFileSync(USERSCRIPT_HEADER0, "utf-8")

  let config = {
    package: packagejson,
    config: packagejson.config,
    addon: { definitions: addonHeader }
  }

  let parsedUserscriptHeader = parseUserscriptHeader(userscriptHeader, config)

  config["userscript"] = {
    header: parsedUserscriptHeader[0],
    definitions: parsedUserscriptHeader[1]
  }

  manifest = generateManifest(manifestSkel, config)
  let addon = generateAddon(jsSkel, config)
  let userscript = generateUserscript(jsSkel, config)

  // save files
  console.log("Saving file: " + ADDON_JS0)
  fs.writeFileSync(ADDON_JS0, addon)
  console.log("Saving file: " + MANIFEST0)
  fs.writeFileSync(MANIFEST0, manifest)
  console.log("Saving file: " + USERSCRIPT0)
  fs.writeFileSync(USERSCRIPT0, userscript)

  // generate zip archive
  let zip = new JSZip()
    .file(strippedFileName(MANIFEST0), manifest)
    .file(strippedFileName(ADDON_JS0), addon)
    .file(strippedFileName(ICON0), fs.readFileSync(ICON0))

  // save zip archive
  const archive0 = ARCHIVE0.replace("{}", config.package.version)
  zip.generateNodeStream({ type: "nodebuffer", streamFiles: true })
    .pipe(fs.createWriteStream(archive0))
    .on("finish", () => {
      console.log("Created archive: " + archive0)
      console.log("Deployed version " + config.package.version + " sucessfully!")
    })

}

if (require.main === module) 
  main()

